src/Microsoft.DotNet.CodeFormatting/Rules/MarkReadonlyFieldsRule.cs

```diff
@@ -0,0 +1,527 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.DotNet.CodeFormatting.Rules
+{
+    /// <summary>
+    /// Mark any fields that can provably be marked as readonly.
+    /// </summary>
+    [GlobalSemanticRuleOrder(GlobalSemanticRuleOrder.MarkReadonlyFieldsRule)]
+    internal sealed class MarkReadonlyFieldsRule : IGlobalSemanticFormattingRule
+    {
+        private readonly SemaphoreSlim _processUsagesLock = new SemaphoreSlim(1, 1);
+        private ConcurrentDictionary<IFieldSymbol, bool> _unwrittenWritableFields;
+
+        public bool SupportsLanguage(string languageName)
+        {
+            return languageName == LanguageNames.CSharp;
+        }
+
+        public async Task<Solution> ProcessAsync(
+            Document document,
+            SyntaxNode syntaxRoot,
+            CancellationToken cancellationToken)
+        {
+            if (_unwrittenWritableFields == null)
+            {
+                using (await SemaphoreLock.GetAsync(_processUsagesLock))
```

Comment:
This part makes me nervous.  The code formatter is optimized for two specific scenarios:

1. For conforming code do as little as possible.
2. For non-conforming code clean up the code as fast as possible

The first issue has higher priority than the second.  The logic being we want the tool to be fast enough that it could be run essentially on build.  

This particular check though does seem fairly expensive as you noted.  I'm worried about the perf impact it's going to have on clean runs of the tool. 

Other rules like the private field naming have the same problem.  It's mitigated by delaying the expensive checks by first doing a lot of inexpensive ones.  I'm trying to brainstorm ways that the same could be done here. 

Going to run this on Roslyn later on to get a sense of the true impact.  Maybe it's not that bad and I'm being over paranoid.